packages/vite/src/node/ssr/runnerImport.ts

```diff
@@ -0,0 +1,76 @@
+import type { InlineConfig } from '../config'
+import { resolveConfig } from '../config'
+import { createRunnableDevEnvironment } from '../server/environments/runnableEnvironment'
+import { mergeConfig } from '../utils'
+
+interface RunnerImportResult<T> {
+  module: T
+  dependencies: string[]
+}
+
+/**
+ * Import any file using the default Vite environment.
+ * @experimental
+ */
+export async function runnerImport<T>(
+  moduleId: string,
+  inlineConfig?: InlineConfig,
+): Promise<RunnerImportResult<T>> {
+  const isModuleSyncConditionEnabled = (await import('#module-sync-enabled'))
+    .default
+  const config = await resolveConfig(
+    mergeConfig(inlineConfig || {}, {
+      configFile: false,
+      envFile: false,
+      cacheDir: process.cwd(),
+      environments: {
+        inline: {
+          consumer: 'server',
+          dev: {
+            moduleRunnerTransform: true,
+          },
+          resolve: {
+            external: true,
+            mainFields: [],
+            conditions: [
+              'node',
+              ...(isModuleSyncConditionEnabled ? ['module-sync'] : []),
+            ],
+          },
+        },
+      },
+    } satisfies InlineConfig),
+    'serve',
+  )
+  const environment = createRunnableDevEnvironment('inline', config, {
+    runnerOptions: {
+      hmr: {
+        logger: false,
+      },
+    },
+    hot: false,
+  })
+  await environment.init()
+  try {
+    const module = await environment.runner.import(moduleId)
+    const modules = [
+      ...environment.runner.evaluatedModules.urlToIdModuleMap.values(),
+    ]
+    const dependencies = modules
+      .filter((m) => {
+        // ignore all externalized modules
+        if (!m.meta || 'externalize' in m.meta) {
+          return false
+        }
+        // ignore the current module
+        return m.exports !== module
+      })
+      .map((m) => m.file)
+    return {
+      module,
+      dependencies,
+    }
+  } finally {
+    await environment.close()
+  }
```

Comment:
If we close the environment here, dynamic import happening after the main module import throws `Error: Vite module runner has been closed.`. Here is a repro https://stackblitz.com/edit/93byylkl?file=index.js

Do we need to give up explicit close here and rely on proper gc instead?

Reply:
The environment also overrides global `Error.captureStackTrace` and `close` removes the override, so we would have to either disable this and patch the error if it happens or return the environment and document that it should be closed manually (or just return the `close` method) 

> dynamic import doesn't appear as `dependencies`

Dynamic import is not in dependencies because it wasn't executed. We can store `ssrResult.dependencies` as `module.parsedDependencies` maybe 🤷🏻